src/index.js

```diff
@@ -1,130 +1,236 @@
-var defaultSpecs = [ '**/*.spec.js' ];
-var defaultWatchPaths = [ './src/**/*', './spec/**/*', './resource/**/*' ];
-var processesDefined = false;
-
-process.on( 'uncaughtError', console.log );
-
-function cover( specPath, show ) {
-	var args = Array.prototype.slice.call( arguments, 2 ) || [];
-	args.unshift( specPath, true );
-	return function( /* cb */ ) {
-		return runSpecs.apply( undefined, args )
-			.on( 'end', function() {
-				if ( show ) {
-					open( './coverage/lcov-report/index.html' );
-					process.exit();
-				}
-			} );
-	};
-}
+var _ = require( "lodash" );
+var when = require( "when" );
+var open = require( "open" );
+var path = require( "path" );
+var mocha = require( "gulp-spawn-mocha" );
+var processhost = require( "processhost" )();
+var cmdPostfix = process.platform === "win32" ? ".cmd" : "";
+var defaults = {
+	allSpecPaths: [ "./spec" ],
+	behaviorSpecPaths: [ "./spec/behavior" ],
+	integrationSpecPaths: [ "./spec/integration" ],
+	specs: [ "**/*.spec.js" ],
+	watchPaths: [ "./src/**/*", "./spec/**/*", "./resource/**/*" ],
+	sourcePaths: [ "./src", "./resource" ],
+	sources: [ "**/*.js" ],
+	coverageHtml: "./coverage/lcov-report/index.html",
+	exclude: [ "!node_modules*" ],
+	jscsCfgPath: ".jscsrc"
+};
+var jshint = require( "gulp-jshint" );
+var jscs = require( "gulp-jscs" );
+var gutil = require( "gulp-util" );
+var gulpChanged = require( "gulp-changed" );
+var stylish = require( "jshint-stylish" );
 
-function groklate( ngrok, subdomain, port, token ) {
-	return when.promise( function( resolve, reject ) {
-		ngrok.connect( {
-			port: port,
-			subdomain: subdomain,
-			authtoken: token
-		}, function( err, url ) {
-				if ( err ) {
-					reject( err );
-				} else {
-					resolve( url );
-				}
-			} );
-	} );
+function permuPath( dirs, globs ) {
+	return _.reduce( globs, function( accum, glb ) {
+		Array.prototype.push.apply(
+			accum,
+			_.map( dirs, function( dir ) {
+				return path.join( dir, glb );
+			} )
+		);
+		return accum;
+	}, [] );
 }
 
-function init() {
-	gulp.task( 'restartProcesses', function( cb ) {
-		processhost.restart()
-			.then( function() {
-				cb();
-			}, function( err ) {
-					cb( err );
-				} );
-	} );
-}
+module.exports = function( gulpRef, cfg ) {
+	var gulp = gulpRef;
+	var options = _.defaults( cfg || {}, defaults );
+	var processesDefined = false;
 
-function runSpecs( specPath, coverage ) {
-	var args = Array.prototype.slice.call( arguments, 2 );
-	if ( args.length === 0 ) {
-		args = [ defaultSpecs ];
+	function runSpecs( testCfg ) {
+		return gulp.src( testCfg.specs, { read: false } )
+			.pipe( mocha( {
+				R: "spec",
+				istanbul: testCfg.coverage ? { x: testCfg.coverageExclude } : false
+			} ) );
 	}
-	var joinArgs = _.flatten( args );
-	var specs = _.reduce( joinArgs, function( acc, spec ) {
-		_.each( specPath, function( p ) {
-			acc.push( path.join( p, spec ) );
+
+	function test( specs ) {
+		var s = _.isArray(specs) ? specs: [ specs ];
```

Comment:
This is one of a bunch of places where it looks like we're missing horizontal white space. Maybe run `jscs --fix` again?